File: apps/api/src/modules/bookings/interfaces/api/bookings.controller.ts
import { Body, Controller, Param, Post, Req, UseGuards, UsePipes } from '@nestjs/common';
import { UserRole } from '@/common/enums';
import { z } from 'zod';
import { JwtAuthGuard } from '@/common/guards/jwt-auth.guard';
import { RolesGuard } from '@/common/guards/roles.guard';
import { ZodValidationPipe } from '@/common/validation/zod-validation.pipe';
import { BookingsService } from '../../application/bookings.service';

const createBookingSchema = z.object({
  venueId: z.string().uuid(),
  startsAt: z.string().datetime(),
  endsAt: z.string().datetime()
});

@Controller('bookings')
@UseGuards(JwtAuthGuard)
export class BookingsController {
  constructor(private readonly bookingsService: BookingsService) {}

  @Post()
  @UsePipes(new ZodValidationPipe(createBookingSchema))
  create(@Req() req: Express.Request, @Body() body: z.infer<typeof createBookingSchema>) {
    return this.bookingsService.createBooking(req.user!.sub, {
      venueId: body.venueId,
      startsAt: new Date(body.startsAt),
      endsAt: new Date(body.endsAt)
    });
  }

  @Post(':bookingId/finalize')
  @UseGuards(new RolesGuard([UserRole.ADMIN, UserRole.VENUE_OWNER]))
  finalize(@Param('bookingId') bookingId: string) {
    return this.bookingsService.finalizeBooking(bookingId);
  }
}

File: apps/mobile/lib/features/bookings/presentation/bookings_controller.dart
import 'package:flutter/material.dart';
import 'package:proyecto_fut_app/features/bookings/data/bookings_repository.dart';

class BookingsController extends ChangeNotifier {
  BookingsController(this._repository);

  final BookingsRepository _repository;
  bool loading = false;

  Future<void> createBooking({
    required String venueId,
    required DateTime startsAt,
    required DateTime endsAt,
  }) => _guard(() => _repository.createBooking(venueId: venueId, startsAt: startsAt, endsAt: endsAt));

  Future<void> finalizeBooking(String bookingId) => _guard(() => _repository.finalizeBooking(bookingId));

  Future<void> _guard(Future<void> Function() fn) async {
    loading = true;
    notifyListeners();
    try {
      await fn();
    } finally {
      loading = false;
      notifyListeners();
    }
  }
}
